File: array/two_pointers/separate_pointers/0925_long_pressed_name.js
/*
Your friend is typing his name into a keyboard. Sometimes, when typing a character 
c, the key might get long pressed, and the character will be typed 1 or more times.

You examine the typed characters of the keyboard. Return True if it is possible 
that it was your friends name, with some characters (possibly none) being long 
pressed.
*/

// https://leetcode.com/problems/long-pressed-name/

// two pointers
// time complexity O(m + n)
// space complexity O(1)
var isLongPressedName = function (name, typed) {
  let p1 = 0,
    p2 = 0;
  while (p1 < name.length && p2 < typed.length) {
    if (name[p1] !== typed[p2]) return false;
    p1++;
    p2++;
    if (name[p1] !== name[p1 - 1]) {
      while (typed[p2] === typed[p2 - 1]) {
        p2++;
      }
    }
  }
  if (p1 < name.length || p2 < typed.length) return false;
  return true;
};

File: array/two_pointers/separate_pointers/0925_long_pressed_name.py
# Your friend is typing his name into a keyboard. Sometimes, when typing a character c, the key might get long pressed, and the character will be typed 1 or more times.

# You examine the typed characters of the keyboard. Return True if it is possible that it was your friends name, with some characters (possibly none) being long pressed.

# https://leetcode.com/problems/long-pressed-name/

class Solution:
    def isLongPressedName(self, name: str, typed: str) -> bool:
        p1, p2 = 0, 0
        l1, l2 = len(name), len(typed)
        
        # Traverse through both strings
        while p2 < l2:
            # If both characters match, move both pointers forward
            if p1 < l1 and name[p1] == typed[p2]:
                p1 += 1
                p2 += 1
            # If the current `typed` character is the same as the previous one (long press scenario)
            elif p2 > 0 and typed[p2] == typed[p2 - 1]:
                p2 += 1
            # Mismatch case: Characters don't match and it's not a valid long press
            else:
                return False
        
        # All characters in `name` should be used
        return p1 == l1